sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesInLeftSemiAntiJoin.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{AGGREGATE, LEFT_SEMI_OR_ANTI_JOIN}
+
+/**
+ * Remove the redundant aggregation from left semi/anti join if the same aggregation has already
+ * been done on left side.
+ */
+object RemoveRedundantAggregatesInLeftSemiAntiJoin extends Rule[LogicalPlan] {
```

Comment:
It seems to me that we should have a way to propagate up the distinct key information, and remove unnecessary agg operators.

I can imagine that we may allow more operators between redundant aggregates, left join this time, and probably project, filter, etc. next time.

Reply:
How about just add this rule to `RemoveRedundantAggregates`?
```patch
Index: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala	(revision 86a828aa4e27676ec74dab28b1fb3db996f9c637)
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala	(date 1626706341253)
@@ -20,7 +20,8 @@
 import org.apache.spark.sql.catalyst.analysis.PullOutNondeterministic
 import org.apache.spark.sql.catalyst.expressions.{AliasHelper, AttributeSet}
 import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
-import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan}
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Join, LogicalPlan}
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.catalyst.trees.TreePattern.AGGREGATE
 
@@ -29,7 +30,7 @@
  * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
  */
 object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
-  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithPruning(
     _.containsPattern(AGGREGATE), ruleId) {
     case upper @ Aggregate(_, _, lower: Aggregate) if isLowerRedundant(upper, lower) =>
       val aliasMap = getAliasMap(lower)
@@ -47,6 +48,13 @@
       } else {
         newAggregate
       }
+
+    case agg @ Aggregate(groupingExps, aggExps,
+      j @ Join(left: Aggregate, _, LeftSemi | LeftAnti, _, _)) if agg.groupOnly && left.groupOnly &&
+        aggExps.forall(e => left.aggregateExpressions.exists(_.semanticEquals(e))) &&
+        groupingExps.length == left.groupingExpressions.length &&
+        groupingExps.zip(left.groupingExpressions).forall(e => e._1.semanticEquals(e._2)) =>
+      j
   }
 
   private def isLowerRedundant(upper: Aggregate, lower: Aggregate): Boolean = {

```

Reply:
Can we follow `LogicalPlanStats` and add `trait DistinctKeyVisitor extends LogicalPlanVisitor[Set[ExpressionSet]] { self: LogicalPlan =>` to properly propagate the distinct key info? We can start with a few nodes like left join.